Rework cpu affinity parsing

keszybz opened the pull request:
This is yet-another approach to fix CPUAffinity.

It does **not** fix the endianess issues. @msekletar you wanted to work on this, so I left that part out. Maybe you could simply push your patches to my branch?

There's also a FIXME added. I'll prepare a patch for that later if the general approach in this PR is approved. I don't think it's very important to fix it right now.

poettering commented:
lgtm, modulo the fixme of course

keszybz commented:
Updated to fix the FIXME. Some new patches on top to sanitize how we handle arguments during reload. I'm not entirely happy with this, because we reset things, but don't reinitialize everything fully. Nevertheless, I think those changes make things more predictable.

poettering commented on src/core/main.c:
```diff
@@ -139,6 +139,11 @@ static sd_id128_t arg_machine_id = {};
+static struct CPUAffinity {
+        cpu_set_t *set;
+        size_t allocated;
+} arg_cpu_affinity = {};
```
one of those days we should probably move the structure definition into process-util.h or so...

poettering commented:
```
Direct leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7fee4eb3e720 in __interceptor_realloc (/usr/lib/x86_64-linux-gnu/libasan.so.5+0xe9720)
    #1 0x7fee4dc1f9ab in greedy_realloc ../src/basic/alloc-util.c:63
    #2 0x7fee4dad877a in cpu_set_to_string ../src/shared/cpu-set-util.c:26
    #3 0x557aa0b173ee in test_parse_cpu_set_extend ../src/test/test-cpu-set-util.c:171
    #4 0x557aa0b180cb in main ../src/test/test-cpu-set-util.c:195
    #5 0x7fee4c67009a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2409a)
```

msan doesn't appear to like this one!

poettering commented:
(looks good to me otherwise)

keszybz commented:
Duh, it was a stupid memleak in the test. Fixed now.
I also moved `struct CPUAffinity` to process-util.h as suggested.

poettering commented on src/core/main.c:
```diff
@@ -2184,29 +2203,74 @@ static int do_queue_default_job(
+        arg_default_oom_policy = OOM_STOP;
+
+        arg_cpu_affinity.set = cpu_set_mfree(arg_cpu_affinity.set);
+        arg_cpu_affinity.allocated = 0;
```
maybe move these to lines into a common destructor for the new struct, as part of proces-util.h too

keszybz replied:
Thanks for this. I should rework the code to use the new structure everywhere. This should make the code a bit cleaner.

keszybz commented:
Updated and refreshed.

I rewrote everything to use a `struct CPUSet { cpu_set_t *set, size_allocated }` and pass a pointer to that around, instead of two pointers. This is a bit nicer because the allocation details are encapsulated.

I also tooks @msekletar's patches and rebased them to use CPUSet and added some tests and cleanups.

Michal implemented my idea to use ranges for CPUAffinity= output in `systemctl show`, despite not being convinced that this is the best approach ;). I'm not fully convinced it is either. Independent opinions would be very useful. I think it makes the output readable for humans, but Michal raises the point that this is harder to script, because a list of numbers can be easily imported as a bash array.

PTAL.

msekletar commented:
@keszybz Thanks for porting my patches to CPUSet! Much appreciated. 

Btw, should I close PR #12644 ?

msekletar commented:
Also here is NEWS bullet point text that I intended to add. @keszybz you should probably add it as a final commit to this branch.

```
        * Wire format for CPUAffinity service property has changed. Previously,
          we transferred affinity mask as raw cpu_set_t which was not
          endian-safe. Now, we expect data to be in little-endian byte order.
          To create expected data representation you should iterate over the
          cpu_set_t and create a raw byte array manually instead of
          byte-swapping individual array members of cpu_set_t to avoid issues
          with internal bit width of libc storage type. For inspiration see
          function cpu_set_to_dbus().
```

Feel free to reword it as you see fit.

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -62,24 +63,74 @@ cpu_set_t* cpu_set_malloc(unsigned *ncpus) {
+
+        assert(cpu_set);
+
+        need = CPU_ALLOC_SIZE(ncpus);
```
hmm, maybe use this instead:

```c
need = CPU_ALLOC_SIZE(MAX(ncpus, get_nprocs_conf()));
```

i.e. it might be a good idea to start out with the number of cpus in the system here

keszybz replied:
If we are on a large machine, we might do something like `CPUAffinity=0-3`, and then we are fine with the minimum allocation. I'd just leave things simple here.

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -206,3 +206,42 @@ int cpus_in_affinity_mask(void) {
+        if (!out)
+                return -ENOMEM;
+
+        for (unsigned cpu = 0; cpu < set->allocated * 8; cpu++)
```
unsigned → size_t?

keszybz replied:
It's used in the glibc API which only takes unsigned ...

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -206,3 +206,42 @@ int cpus_in_affinity_mask(void) {
+        assert(bits);
+        assert(set);
+
+        for (unsigned cpu = size * 8; cpu > 0; cpu--)
```
size_t?

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -206,3 +206,42 @@ int cpus_in_affinity_mask(void) {
+
+        for (unsigned cpu = size * 8; cpu > 0; cpu--)
+                if (bits[(cpu - 1) / 8] & (1u << ((cpu - 1) % 8))) {
+                        int r;
```
not that it matters at all, but usually we declare "int r" is the outermost scope

keszybz replied:
Ah, right. I just made a very similar comment in another PR ;)

poettering commented on src/test/test-cpu-set-util.c:
```diff
@@ -179,6 +179,30 @@ static void test_parse_cpu_set_extend(void) {
+        assert_se(cpu_set_from_dbus(array, allocated, &c2) == 0);
+        assert_se(c2.set);
+        assert_se(c2.allocated == c.allocated);
+        assert_se(memcmp(c.set, c2.set, c.allocated) == 0);
```
maybe include a complex serialization like this also as a binary blob in the test/ directory, then read it each test and make sure it matches what it should? that way we might catch endianess and word size issues if our test cases are every run on weird archs

keszybz replied:
I'll just a an explicit test in `test-cpu-set-util`.

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -34,6 +34,46 @@ char* cpu_set_to_string(const CPUSet *a) {
         return TAKE_PTR(str) ?: strdup("");
 }
 
+char *cpu_set_to_range(const CPUSet *set) {
```
i would probably call this  cpu_set_to_range_string()

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -34,6 +34,46 @@ char* cpu_set_to_string(const CPUSet *a) {
+                                range_start = range_end = i;
+                                in_range = true;
+                        }
+                } else if (!b && in_range) {
```
the `!b &&` part is redundant, given that this is the second branch and the first branch already covered the `b` case

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -34,6 +34,46 @@ char* cpu_set_to_string(const CPUSet *a) {
+        int r;
+
+        for (unsigned i = 0; i < set->allocated * 8; i++) {
+                bool b = CPU_ISSET_S(i, set->allocated, set->set);
```
I figure this variable can go, and the `CPU_ISSET_S()` expression be moved directly into the `if` test

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -34,6 +34,46 @@ char* cpu_set_to_string(const CPUSet *a) {
+                        if (!GREEDY_REALLOC(str, allocated, len + 2 + 2 * DECIMAL_STR_MAX(unsigned)))
+                                return NULL;
+
+                        r = sprintf(str + len, len > 0 ? " %d-%d" : "%d-%d", range_start, range_end);
```
hmm, this would mean if only cpu 5 is marked we'd output `5-5` for it, no? I think it would be nicer to output just `5` in that case, i.e. handle the `range_start == range_end` case special

poettering replied:
i see this has been fixed in a later commit in the same PR, so ignore me on this one

poettering commented:
> Also here is NEWS bullet point text that I intended to add. @keszybz you should probably add it as a final commit to this branch.
> 
> ```
>         * Wire format for CPUAffinity service property has changed. Previously,
>           we transferred affinity mask as raw cpu_set_t which was not
>           endian-safe. Now, we expect data to be in little-endian byte order.
>           To create expected data representation you should iterate over the
>           cpu_set_t and create a raw byte array manually instead of
>           byte-swapping individual array members of cpu_set_t to avoid issues
>           with internal bit width of libc storage type. For inspiration see
>           function cpu_set_to_dbus().
> ```
> 
> Feel free to reword it as you see fit.

I would probably just consider this a bugfix (which we usually don't document in NEWS), and it won't affect LE machines anyway, i.e. only exotic cpus

msekletar commented:
> I would probably just consider this a bugfix (which we usually don't document in NEWS), and it won't affect LE machines anyway, i.e. only exotic cpus

Fair enough, however, 32bit LE clients will be affected as well, i.e. they will get different serialisation if internal storage type is 32bit in width.

keszybz commented on src/core/dbus-execute.c:
```diff
@@ -1566,37 +1566,22 @@ int bus_exec_context_set_transient_property(
                         } else {
                                 _cleanup_free_ char *str = NULL;
-                                size_t ncpus;
+                                const CPUSet set = {(cpu_set_t*) a, n};
```
It's removed in subsequent patch.

keszybz commented on src/shared/cpu-set-util.c:
```diff
@@ -62,24 +63,74 @@ cpu_set_t* cpu_set_malloc(unsigned *ncpus) {
 }
 
-int parse_cpu_set_internal(
+static int cpu_set_realloc(CPUSet *cpu_set, unsigned ncpus) {
```
I don't think so. The glibc interface uses unsigned to number the cpus, so I think we should stick to this. Even if we allocated a bigger array, we wouldn't be able to index it anyway.

keszybz commented:
> 32bit LE clients will be affected as well, i.e. they will get different serialisation if internal storage type is 32bit in width.

I don't think so.
For 64 bit: [byte0=bits-0-7, byte1=bits 8-15, ..., byte7=bits 56-63]
For 32 bit: [byte0=bits-0-7, byte1=bits 8-15, ..., byte7=bits 56-63]
So the only difference is that the length might be different in the two cases. But I don't think we made promises about length.

Updated:
- cpu_set_to_range is renamed to cpu_set_to_range_string
- the suggested code improvements are done, except as noted above

I'll add the NEWS entry later, because right now it'd conflict with other PRs. I think we should mention the wire-format change and the systemctl show change, but without going into any coding details.
